tensorflow/contrib/avro/python/tests/parse_avro_record_test.py

```diff
@@ -0,0 +1,1143 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+import logging
+import collections
+
+from tensorflow.python.framework import dtypes as tf_types
+from tensorflow.python.framework.errors import OpError
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import parsing_ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.platform import test
+
+from tensorflow.contrib.avro.python.parse_avro_record import parse_avro_record
+from tensorflow.contrib.avro.python.utils.avro_record_utilities import get_record_value, serialize, parse_schema
+from tensorflow.contrib.avro.python.utils.tensor_utilities import fill_in_fixed_len, fill_in_fixed_len_sequence, \
+    get_end_indices, get_n_elements_per_batch
+import tensorflow.contrib.avro.python.utils.numerr as nr
+
+
+class DataForTest(collections.namedtuple('DataForTest', ['schema', 'data', 'features', 'should_pass'])):
+    """
+    Test data contains a schema, data, features (for parsing), and whether this test is expected to pass or fail.
+    """
+
+
+class ParseAvroRecordTest(test.TestCase):
+
+    def __init__(self, *args, **kwargs):
+        super(ParseAvroRecordTest, self).__init__(*args, **kwargs)
+
+    def setUp(self):
+        """
+        Setup fixture for test cases.
+        """
+        log_root = logging.getLogger()  # set logging level
+        log_root.setLevel(logging.INFO)
+
+    def run_test(self, test_case):
+        """
+        Runs a test case.
+        """
+        # Run over all test cases
+        schema_object = parse_schema(test_case.schema)
+        with ops.Graph().as_default() as g, self.test_session(graph=g) as sess:
+            str_input = array_ops.placeholder(tf_types.string)
+
+            parsed = parse_avro_record(str_input, test_case.schema, test_case.features)
+            records = test_case.data
+            # To test batch processing summarize all test data points into one batch
+            serialized = [serialize(record, schema_object) for record in records]
+            # If this test case should pass ensure that all thresholds are met
+            if test_case.should_pass:
+                # Here is where we execute our parser within TensorFlow
+                tensors = sess.run(parsed, feed_dict={str_input: serialized})
+                # Go over all key, value pairs in the features; keys are strings and values are TensorFlow type info
+                for key, value in test_case.features.iteritems():
+                    # Get all intended tensor values from the test data
+                    tensor_be = len(records)*[None]
+                    for i_record, record in enumerate(records):
+                        record_value = get_record_value(record, key, value.dtype)
```

Comment:
get_record_value is non-trivial, so it's hard to know what this is returning and probably needs test cases itself. Instead of automatically deriving the expected values here, I would add the expected values to your individual test cases below. This is more verbose, but will make the tests much easier to read and more obviously correct.

Reply:
Good point. I've changed the design of these test cases following your suggestion.